File: src/GS.Certifications.Domain/Entities/Comprobantes/Comprobante.cs
using GS.Certifications.Domain.Entities.Empresas;
using GS.Certifications.Domain.Entities.Impuestos;
using GS.Certifications.Domain.Entities.Percepciones;
using GS.Certifications.Domain.Entities.Periodos;
using GSF.Domain.Common;
using GSF.Domain.Entities.Companies;
using GSF.Domain.Entities.Global;
using System;
using System.Collections.Generic;

namespace GS.Certifications.Domain.Entities.Comprobantes;

public class Comprobante : BaseIntEntity
{
    public Guid Guid { get; private set; }
    public string NroIdentificacionFiscalPro { get; set; }
    public string DomicilioPro { get; set; }
    public short? CategoriaTipoEmisorId { get; set; }
    public CategoriaTipo CategoriaTipoEmisor { get; set; }
    public short? ComprobanteTipoId { get; set; }
    public ComprobanteTipo ComprobanteTipo { get; set; }
    // TODO: no es nullable
    public short? CategoriaTipoReceptorId { get; set; }
    public CategoriaTipo CategoriaTipoReceptor { get; set; }
    public string NroIdentificacionFiscalCompany { get; set; }
    public int PuntoVenta { get; set; }
    public string Letra { get; set; }
    public int Numero { get; set; }
    public DateTime? FechaEmision { get; set; }
    public DateTime? FechaVencimiento { get; set; }
    public DateTime? FechaVencimientoCodigoAutorizacion { get; set; }
    public DateTime FechaRegistracion { get; set; }
    public short? TipoCodigoAutorizacionId { get; set; }
    public CodigoAutorizacionTipo TipoCodigoAutorizacion { get; set; }
    public string CodigoAutorizacion { get; set; }
    public long? MonedaId { get; set; }
    public Currency Moneda { get; set; }
    public decimal ImporteNeto { get; set; } = default;
    public decimal ImporteTotal { get; set; } = default;
    public decimal ImporteIVA { get; set; } = default;
    public decimal ImporteBonificacion { get; set; } = default;
    public decimal ImportePercepcionIVA { get; set; } = default;
    public decimal ImportePercepcionIIBB { get; set; } = default;
    public decimal ImportePercepcionMunicipal { get; set; } = default;
    public decimal ImporteImpuestosInternos { get; set; } = default;
    public decimal ImporteOtrosTributosProv { get; set; } = default;
    // TODO: no es nullable
    public int? EmpresaId { get; set; }
    public EmpresaPortal Empresa { get; set; }
    // TODO: no es nullable
    public long? CompanyId { get; set; }
    public Company Company { get; set; }
    public string CodigoHES { get; set; }
    public int ComprobanteEstadoId { get; set; } = ComprobanteEstado.BORRADOR;
    public ComprobanteEstado ComprobanteEstado { get; set; }
    public string Comentarios { get; set; }
    public string MotivoRechazo { get; set; }
    public DateTime? FechaRechazo { get; set; }
    public string UsuarioRechazo { get; set; }
    public List<ComprobanteDetalle> Detalles { get; set; } = new();
    public List<ImpuestoDetalle> Impuestos { get; set; } = new();
    public List<PercepcionDetalle> Percepciones { get; set; } = new();

    public bool? ValidacionQR { get; set; } = false;
    public string QRValor { get; set; }

    public short? EstadoValidacionARCAQRId { get; set; } = EstadoValidacionARCA.NO_VALIDADA;
    public EstadoValidacionARCA EstadoValidacionARCAQR { get; set; }

    public short? EstadoValidacionARCAId { get; set; } = EstadoValidacionARCA.NO_VALIDADA;
    public EstadoValidacionARCA EstadoValidacionARCA { get; set; }
    public string ObservacionesARCA { get; set; }
    public string ObservacionesARCAQR { get; set; }

    public Periodo Periodo { get; set; }
    public int? PeriodoId { get; set; }

    public short? OrigenId { get; set; }
    public Origen Origen { get; set; }
    public short? PropietarioActualId { get; set; }
    public Origen PropietarioActual { get; set; }
    public short? CondicionVentaId { get; set; } = CondicionVenta.OTRO;
    public CondicionVenta CondicionVenta { get; set; }
    public string NroRemito { get; set; }
    public string NroOrdenCompra { get; set; }
    public decimal? Cotizacion { get; set; }
    public string NombreArchivo { get; set; }

    public Comprobante()
    {
        Guid = Guid.NewGuid();
    }
}


public class ComprobanteDetalle : BaseIntEntity
{
    public int ComprobanteId { get; set; }
    public Comprobante Comprobante { get; set; }
    public short? UnidadMedidaId { get; set; }
    public UnidadMedida UnidadMedida { get; set; }
    public int Cantidad { get; set; }
    public decimal PrecioUnitario { get; set; } = default;
    public decimal ImporteBonificacion { get; set; } = default;
    public decimal Subtotal { get; set; } = default;
    public decimal ImporteIVA { get; set; } = default;
    public bool Exento { get; set; }
    public decimal Alicuota { get; set; } = default;
    public int? OrdenCompraId { get; set; }
    public string Detalle { get; set; }
}

File: src/GS.Certifications.Web/wwwroot/src/Proveedores/Comprobantes/Comprobante.js
import ComprobanteDetalle from "./ComprobanteDetalle";
import ComprobantePercepcion from "./ComprobantePercepcion";
import ComprobanteImpuesto from "./ComprobanteImpuesto";

const NO_VALIDADA = 4;

export default class Comprobante {
    constructor(dto) {
        this.id = dto ? dto.id : null;
        this.guid = dto ? dto.guid : null;
        this.rowVersion = dto ? dto.rowVersion : null;
        this.nroIdentificacionFiscalPro = dto ? dto.nroIdentificacionFiscalPro : null;
        this.domicilioPro = dto ? dto.domicilioPro : null;
        this.categoriaTipoEmisorId = dto ? dto.categoriaTipoEmisorId : null;
        this.comprobanteTipoId = dto ? dto.comprobanteTipoId : null;
        this.categoriaTipoReceptorId = dto ? dto.categoriaTipoReceptorId : null;
        this.nroIdentificacionFiscalCompany = dto ? dto.nroIdentificacionFiscalCompany : null;
        this.puntoVenta = dto ? dto.puntoVenta : null;
        this.letra = dto ? dto.letra : null;
        this.numero = dto ? dto.numero : null;
        this.fechaEmision = dto ? dto.fechaEmision ? new Date(dto.fechaEmision).toISOString().split('T')[0] : null : null;
        this.fechaRegistracion = dto ? dto.fechaRegistracion : null;
        this.tipoCodigoAutorizacionId = dto ? dto.tipoCodigoAutorizacionId : null;
        this.codigoAutorizacion = dto ? dto.codigoAutorizacion : null;
        this.monedaId = dto ? dto.monedaId : null;
        this.importeNeto = dto ? dto.importeNeto ? dto.importeNeto : 0 : 0;
        this.importeTotal = dto ? dto.importeTotal ? dto.importeTotal : 0 : 0;
        this.importeIVA = dto ? dto.importeIVA ? dto.importeIVA : 0 : 0;
        this.importeBonificacion = dto ? dto.importeBonificacion ? dto.importeBonificacion : 0 : 0;
        this.iva21 = dto ? dto.iva21 ? dto.iva21 : 0 : 0;
        this.iva105 = dto ? dto.iva105 ? dto.iva105 : 0 : 0;
        this.empresaId = dto ? dto.empresaId : null;
        this.companyId = dto ? dto.companyId : null;
        this.codigoHES = dto ? dto.codigoHES : null;
        this.comprobanteEstadoId = dto ? dto.comprobanteEstadoId : null;
        this.comentarios = dto ? dto.comentarios : null;

        this.importeImpuestosInternos = dto ? dto.importeImpuestosInternos : 0;
        this.importeOtrosTributosProv = dto ? dto.importeOtrosTributosProv : 0;
        this.importePercepcionIVA = dto ? dto.importePercepcionIVA : 0;
        this.importePercepcionIIBB = dto ? dto.importePercepcionIIBB : 0;

        this.motivoRechazo = dto ? dto.motivoRechazo ? dto.motivoRechazo : "" : "";
        this.proveedor = dto ? dto.proveedor ? dto.proveedor : "" : "";

        this.detalles = [];
        this.impuestos = [];
        this.percepciones = [];

        this.validacionQR = dto ? dto.validacionQR : false;
        this.qrValor = dto ? dto.qrValor : null;
        this.periodoId = dto ? dto.periodoId : null;

        this.estadoConstatacion = dto ? dto.estadoConstatacion : NO_VALIDADA;
        this.estadoValidacionARCAId = dto ? dto.estadoValidacionARCAId : null;
        this.estadoValidacionARCAQRId = dto ? dto.estadoValidacionARCAQRId : null;
        this.observacionesARCA = dto ? dto.observacionesARCA : null;

        this.condicionVentaId = dto ? dto.condicionVentaId : null;
        this.nroRemito = dto ? dto.nroRemito : null;
        this.nroOrdenCompra = dto ? dto.nroOrdenCompra : null;
        this.cotizacion = dto ? dto.cotizacion : 1;

        this.propietarioActualId = dto ? dto.propietarioActualId : null;

        this.nombreArchivo = dto ? dto.nombreArchivo : "";

        this.comprobante = dto ? dto.comprobante : "";
        
        this.fechaVencimiento = dto ? dto.fechaVencimiento ? new Date(dto.fechaVencimiento).toISOString().split('T')[0] : null : null;
        this.fechaVencimientoCodigoAutorizacion = dto ? dto.fechaVencimientoCodigoAutorizacion ? new Date(dto.fechaVencimientoCodigoAutorizacion).toISOString().split('T')[0] : null : null;

        if (dto && dto.detalles && dto.detalles.length) {
            dto.detalles.forEach(element => {
                this.detalles.push(new ComprobanteDetalle(element));
            });
        }

        if (dto && dto.impuestos && dto.impuestos.length) {
            dto.impuestos.forEach(element => {
                this.impuestos.push(new ComprobanteImpuesto(element));
            });
        }

        if (dto && dto.percepciones && dto.percepciones.length) {
            dto.percepciones.forEach(element => {
                this.percepciones.push(new ComprobantePercepcion(element));
            });
        }
    }
}